### PR TITLE
Adds `hasState` optional arg and renames useEntry to useState

### PR DESCRIPTION
**Pull request by chadwatson**
For optimal flexibility this change adds an optional argument to test if the resource has state when trying to `use` it. 99% of use cases won't need to use it, but there are some weird edge cases where it is useful.

While I was at it I was finding `useEntry` to be a bit of a confusing name. I felt like `useState` actually made a little more sense. Since we would need to do **another** major release I went ahead and renamed it to see what you guys thought.

I also took a stab at improving the the docs for `useState`.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -267,59 +272,150 @@ const UserProfile = ({ userId }) => (
...
+          </article>
+        </li>
+      ))}
+    </ul>
+  );
+};
+```
 
-  useAutoSave(post, savePost);
+When `UserTweets` renders, the `useState` hook will take the given `userId`, check if tweets have been fetched for this user, and if they have they will be returned as the first item in the tuple. If they haven't then the loader function will be called and the returned promise will be thrown, which will be caught by the nearest `RemoteResourceBoundary`. Whatever data the promise resolves with will be set as the tweets for that user. Now that the data has been received for this user the component will re-render with the new data. If the component re-renders after this the loader function will not be called since the tweets have already been fetched. We will always get back the latest state.
```
Comment (smacpherson64) [inline README.md]:
```suggestion
When `UserTweets` renders, the `useState` hook will take the given `userId`, and check if tweets have been fetched for this user. If they have, they will be returned as the first item in the tuple. If they haven't then the loader function will be called and the returned promise will be thrown, which will be caught by the nearest `RemoteResourceBoundary`. Whatever data the promise resolves with will be set as the tweets for that user. Once the data has been received for this user, the state will be set and the component will re-render with the new state. If the component re-renders after this the loader function will not be called since the tweets have already been fetched. We will always get back the latest state.
```

File: README.md [hunk 2]
```diff
@@ -267,59 +272,150 @@ const UserProfile = ({ userId }) => (
...
+Take the following resource:
+
+```javascript
+const tweetsResource = createKeyedResource(
+  userId => userId,
+  userId => fetch(`/api/users/${userId}/tweets`).then(res => res.json())
+);
+```
+
+This resource can fetch and store tweets for multiple users. Once it has fetched the tweets for a user it won't re-fetch them. Let's create a component that will render a user's tweets.
```
Comment (smacpherson64) [inline README.md]:
```suggestion
This resource fetches and stores tweets organized by userId. Once it has fetched the tweets for a user it won't re-fetch them. Let's create a component that will render a user's tweets.
```

File: README.md [hunk 3]
```diff
@@ -267,59 +272,150 @@ const UserProfile = ({ userId }) => (
...
+```javascript
+const authTokenResource = createSingleEntryResource(
+  () =>
+    new Promise(resolve => resolve(localStorage.getItem("authToken") || ""))
+);
+
+const profileResource = createSingleEntryResource(authToken =>
+  fetch("/my_profile", {
+    headers: {
+      Authorization: authToken
```
Comment (smacpherson64) [inline README.md]:
```suggestion
      Authorization: `Bearer ${authToken}`
```

File: src/__mocks__/create-mock-resource.js [hunk 4]
```diff
@@ -1,22 +1,24 @@
...
+  selectState,
+  setState,
   loader,
+  hasState,
   entriesExpireAfter
 ) => {
   const spies = {
-    getter: jest.fn().mockImplementation(getter),
-    setter: jest.fn().mockImplementation(setter),
+    selectState: jest.fn().mockImplementation(selectState),
```
Comment (smacpherson64) [inline src/__mocks__/create-mock-resource.js]: `selectState` is a much better name!

File: src/__mocks__/create-mock-resource.js [hunk 5]
```diff
@@ -1,22 +1,24 @@
...
   };
 
   return [
     createResource(
-      spies.getter,
-      spies.setter,
+      spies.selectState,
+      spies.setState,
       spies.loader,
+      hasState,
```
Comment (smacpherson64) [inline src/__mocks__/create-mock-resource.js]: `hasState` is a really good name for this as well!

File: README.md [hunk 6]
```diff
@@ -38,23 +38,22 @@ yarn add react-remote-resource
...
-  10000,
+  userId => fetch(`/api/users/${userId}/tweets`).then(res => res.json()),
+  10000
 );
 
 const UserInfo = ({ userId }) => {
-  const [user] = userResource.useEntry(userId);
-  const [tweets] = tweetsResource.useEntry(userId);
+  const [user] = userResource.useState(userId);
+  const [tweets] = tweetsResource.useState(userId);
```
[the remark this answers is not in the document]
Comment (smacpherson64) [reply README.md]: **Discussion Note**

Should we remove all references to entry? For example `createSingleEntryResource` and the `What is a resource entry?` sections?

[the remark this answers is not in the document]
Comment (chadwatson) [reply README.md]: I wondered about that. What would we name `createSingleEntryResource` then?

**Comment by chadwatson** [comment]
True. Since we're trying to avoid the word "entry" I'm not sure what else to call it. Got any ideas?

File: README.md (hunk 6 above)
[the remark this answers is not in the document]
Comment (smacpherson64) [reply README.md]: Hrm, `createSingleResource` or `createSimpleResource` may work?

**Comment by smacpherson64** [comment]
This is a bit verbose: `expireSelectedStateAfter` possibly: `expireKeyAfter`?

**Comment by smacpherson64** [comment]
Went through it again! Everything looks good just have the discussion on naming! Either way, the direction again is good!

**Comment by chadwatson** [comment]
@smacpherson64 I just pushed up the changes we talked about earlier. I have one more point of discussion while we're still toying with the API.

I'm considering switching the order of the arguments for `createKeyedResource` and making the `createKey` optional. We could default it to a function like `args => args.join("-")` or something. I'd guess that this would probably cover a large amount of the use cases out there.

Thoughts?

**Comment by chadwatson** [comment]
Or maybe we could use [`object-hash`](https://www.npmjs.com/package/object-hash) by default to create the cache keys.

**Comment by chadwatson** [comment]
That last commit it has it in there. Let me know what you think.
